drape_frontend/read_manager.cpp

```diff
@@ -280,18 +280,48 @@ void ReadManager::SetTrafficEnabled(bool trafficEnabled)
   }
 }
 
-void ReadManager::UpdateCustomSymbols(CustomSymbols && symbols)
+void ReadManager::SwapCustomSymbolsContext(CustomSymbolsContextPtr context)
 {
-  auto customSymbolsContext = std::make_shared<CustomSymbolsContext>(std::move(symbols));
-
 #if !defined(OMIM_OS_LINUX)
-  std::atomic_exchange(&m_customSymbolsContext, customSymbolsContext);
+  std::atomic_exchange(&m_customSymbolsContext, context);
 #else
   {
     lock_guard<mutex> guard(m_customSymbolsContextMutex);
-    std::swap(m_customSymbolsContext, customSymbolsContext);
+    std::swap(m_customSymbolsContext, context);
   }
 #endif
 }
 
+void ReadManager::UpdateCustomSymbols(CustomSymbols const & symbols)
+{
+  CustomSymbols currentSymbols = m_customSymbolsContext ? m_customSymbolsContext->m_symbols :
+                                 CustomSymbols();
+  for (auto const & s : symbols)
+    currentSymbols[s.first] = s.second;
+  SwapCustomSymbolsContext(std::make_shared<CustomSymbolsContext>(std::move(currentSymbols)));
+}
+
+void ReadManager::RemoveCustomSymbols(MwmSet::MwmId const & mwmId, std::vector<FeatureID> & leftoverIds)
+{
+  if (!m_customSymbolsContext)
+    return;
+
+  CustomSymbols currentSymbols;
+  leftoverIds.reserve(m_customSymbolsContext->m_symbols.size());
```

Comment:
So, why do we need atomic_exchange, if m_customSymbolContext can be used from only single thread?